### PR TITLE
Implement user sign up

### PR DESCRIPTION
**Pull request by tfidfwastaken**
Closes [sc-110]: As a Guest, I want to sign up for an account, so I can create and share forms

This finishes the user sign up stories. Unfortunately, the diff is a bit bloated because of all the boilerplate needed to set up the database, specs and testing. Since all of those are pretty interconnected, I couldn't find a simple way to split it into smaller, easy to review chunks. Let me know if a synchronous review can speed things up.

The meat of the functionality can be found in 6e3224f8173390856fb044b49823a59b465bad3d.

### CODE REVIEW

File: src/bugle_forms/config.clj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+(ns bugle-forms.config
+  (:require
+   [aero.core :refer [read-config]]
+   [clojure.java.io :as io]))
+
+(defn profile
```
Comment (ssrihari) [inline src/bugle_forms/config.clj]: Consider renaming to env/environment?

File: src/bugle_forms/config.clj [hunk 2]
```diff
@@ -0,0 +1,16 @@
+(ns bugle-forms.config
+  (:require
+   [aero.core :refer [read-config]]
```
Comment (ssrihari) [inline src/bugle_forms/config.clj]: Prefer aliases wherever possible.

File: src/bugle_forms/core.clj [hunk 3]
```diff
@@ -1,29 +1,37 @@
 (ns bugle-forms.core
```
Comment (ssrihari) [inline src/bugle_forms/core.clj]: [nit] "Core" might not be a good name. For anything.

File: src/bugle_forms/core.clj [hunk 4]
```diff
@@ -1,29 +1,37 @@
...
+      wrap-session
+      m/wrap-exception-handler))
 
 (defonce server (atom nil))
 
 (defn start-server []
   (reset! server
           (raj/run-jetty app
-                         {:port (Integer. (or (System/getenv "PORT") "8080"))
+                         {:port (:port (config))
```
Comment (ssrihari) [inline src/bugle_forms/core.clj]: 1. Read config once, story in a read-only state.
2. Look for a simpler API like `(config/lookup :port)`

File: src/bugle_forms/db/connection.clj [hunk 5]
```diff
@@ -0,0 +1,9 @@
+(ns bugle-forms.db.connection
+  (:require
+   [bugle-forms.config :refer [config]]
+   [next.jdbc :as jdbc]))
+
+(def db-spec (:db-spec (config)))
+
+(def datasource
+  (jdbc/get-datasource db-spec))
```
Comment (ssrihari) [inline src/bugle_forms/db/connection.clj]: Remember these are loaded at compile time. Consider using mount to manage component state.

File: src/bugle_forms/db/utils.clj [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+(defn key-exists?
+  "Check if an entry exists in a database.
+  Takes a database handle, the name of the table and column as a kebab-case
+  keyword, and the value whose existence needs to be checked."
+  [ds table key val]
+  (not-empty
+   (sql/find-by-keys ds table {key val}
+                     {:table-fn csk/->snake_case
+                      :column-fn csk/->snake_case})))
```
Comment (ssrihari) [inline src/bugle_forms/db/utils.clj]: 1. Move `table-fn`, `column-fn` out of this function, and perhaps into a config somewhere.
2. `user/find-by-email` might be more domain specific, and appropriate at this time of design.
3. `sql/find-by-keys` is already what you're looking to build.

File: src/bugle_forms/handlers/user.clj [hunk 7]
```diff
@@ -1,9 +1,48 @@
...
+                              "Stub for log-in"))))
 
-(defn signup [_]
-  (response/response "Stub for sign-up"))
+(defn signup
+  "Display signup form."
+  [{:keys [flash]}]
+  (response/response
+   (html5 (layout/application {:title "Sign up" :flash flash}
+                              user-views/signup))))
```
Comment (ssrihari) [inline src/bugle_forms/handlers/user.clj]: View and handler code in same ns?

Comment (ssrihari) [reply src/bugle_forms/handlers/user.clj]: `(response/response view/sign-up-page)`

File: src/bugle_forms/handlers/user.clj [hunk 8]
```diff
@@ -1,9 +1,48 @@
...
+      (dbutil/key-exists? db/datasource :user-account
+                          :email (:user/email account))
+      (util/error-redirect "/signup" "User already exists.")
+
+      :else
+      (do
+        (sql/insert! db/datasource :user-account account
+                     {:table-fn csk/->snake_case})
+        (-> (response/redirect "/login" :see-other)
+            (assoc :flash "Account creation successful!"))))))
```
Comment (ssrihari) [inline src/bugle_forms/handlers/user.clj]: Read about:
- validation (structural, semantic, domain)
- schemas (plumatic/schema swagger, openAPI, etc)

File: src/bugle_forms/routes.clj [hunk 9]
```diff
@@ -1,11 +1,14 @@
...
    [bugle-forms.handlers.core :as core-handlers]
    [bugle-forms.handlers.user :as user-handlers]))
 
 (def routes
   ["/"
    {""       {:get core-handlers/home}
-    "signup" {:get user-handlers/signup}
-    "login" {:get user-handlers/login}
+    "signup" {:get user-handlers/signup
+              :post user-handlers/create-user}
```
Comment (ssrihari) [inline src/bugle_forms/routes.clj]:
```clojure
(def routes
{:post
   {:handler user-handler/create-user
    :schema ""spec/signup-form}
}
})


(def bidi-routes
  (->> routes ...))
```

File: src/bugle_forms/handlers/user.clj (hunk 8 above)
Comment (ssrihari) [reply src/bugle_forms/handlers/user.clj]: Move the actual domain logic into its own ns.

Comment (ssrihari) [reply src/bugle_forms/handlers/user.clj]: Read about MVC.

File: src/bugle_forms/handlers/user.clj [hunk 10]
```diff
@@ -1,9 +1,48 @@
...
+   [bugle-forms.handlers.utils :as util]
+   [bugle-forms.model.user :as user]
+   [bugle-forms.specs :as specs]
+   [bugle-forms.views.user :as user-views]
+   [bugle-forms.views.layout :as layout]
+   [camel-snake-kebab.core :as csk]
+   [clojure.spec.alpha :as s]
+   [hiccup.page :refer [html5]]
+   [next.jdbc.sql :as sql]
    [ring.util.response :as response]))
```
Comment (ssrihari) [inline src/bugle_forms/handlers/user.clj:7]: Use a large list of dependencies as a measure of the concerns in the ns.

File: src/bugle_forms/handlers/user.clj (hunk 8 above)
Comment (ssrihari) [reply src/bugle_forms/handlers/user.clj]: `user/create`

File: resources/migrations/20220209094622-create-user.up.sql [hunk 11]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS user_account (
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
```
Comment (ssrihari) [inline resources/migrations/20220209094622-create-user.up.sql:2]: consider `users`

File: src/bugle_forms/model/user.clj [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+   [buddy.hashers :as hashers]))
+
+(defn signup-params->account
+  "Make a user account suitable to be inserted in database.
+  Takes signup form params and creates a user account with the password hashed
+  using Argon2id."
+  [{:keys [name email password]}]
+  {:user/name name
+   :user/email email
+   :user/password (hashers/derive password {:alg :argon2id})})
```
Comment (ssrihari) [inline src/bugle_forms/model/user.clj:16]: `(def ^:private table-name :user-accounts)`

File: src/bugle_forms/views/user.clj [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+     (util/labelled-text-input
+      "Email Address"
+      :required true
+      :id "email"
+      :name "email"
+      :type "email")]
+    [:div {:id "password"}
+     (util/labelled-text-input
+      "Set a password"
+      :required true
```
Comment (ssrihari) [inline src/bugle_forms/views/user.clj:27]: You might want to derive this information from spec.

File: src/bugle_forms/views/utils.clj [hunk 14]
```diff
@@ -0,0 +1,15 @@
+(ns bugle-forms.views.utils
+  (:require
+   [camel-snake-kebab.core :as csk]))
+
+(defn labelled-text-input
+  "Create a representation of a labelled input box.
+  Takes a label along with an option map which are added as CSS attributes."
+  [label & args]
```
Comment (ssrihari) [inline src/bugle_forms/views/utils.clj]: `[label & {:as args}]`

Comment (ssrihari) [reply src/bugle_forms/views/utils.clj]: `[label options]`

File: .github/workflows/deploy-production.yml [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    - uses: actions/download-artifact@v2
+      with:
+        name: bugle-forms-uberjar
+        path: target/uberjar
+    - name: Setup ssh
+      run: |
+        mkdir -p ~/.ssh/
+        echo "${{ secrets.PROD_PRIVATE_KEY }}" > ~/.ssh/id_rsa
+        sudo chmod 600 ~/.ssh/id_rsa
+        echo "${{ secrets.PROD_KNOWN_HOSTS }}" > ~/.ssh/known_hosts
```
Comment (kirang89) [inline .github/workflows/deploy-production.yml:29]: question: What information is configured here to be written into the `known_hosts` file?

File: README.md [hunk 16]
```diff
@@ -8,17 +8,44 @@ you need to make your analytics, polls or events work.
...
+
+To perform various migration tasks:
+
+```
+lein migrations migrate          # complete pending migrations
+lein migrations rollback         # rollback to previous state
+lein migrations up <ids ...>     # migrate specified ids
+lein migrations down <ids ...>   # rollback specified ids
+lein migrations create <name>    # generate migration files with specified name
+```
```
Comment (kirang89) [inline README.md:48]: suggestion: It'd be a good idea to add a section on Deployment with just enough details for a user to visualise the deployment flow.

File: README.md [hunk 17]
```diff
@@ -8,17 +8,44 @@ you need to make your analytics, polls or events work.
...
```
Comment (kirang89) [inline README.md:49]: suggestion: Consider adding a section on Configuration that summarises how configurations are managed and how they could be changed.

File: deploy.yml [hunk 18]
```diff
@@ -0,0 +1,40 @@
+version: '3.9'
+
+services:
+
+  db:
+    image: postgres:14.1
+    environment:
+      POSTGRES_USER: bugle_prod
+      POSTGRES_DB: bugle_prod
+      POSTGRES_PASSWORD_FILE: /run/secrets/db_password
```
Comment (kirang89) [inline deploy.yml:10]: question: How does this path work to retrieve the db password?

File: deploy.yml [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+    restart: always
+
+volumes:
+  pgdata:
+
+secrets:
+  db_password:
+    file: /home/deploy/.config/bugle-forms/db_password.txt
+  app_secrets:
+    file: /home/deploy/.config/bugle-forms/secrets.edn
```
Comment (kirang89) [inline deploy.yml:40]: question: How are these files created?

**Review by kirang89 — COMMENTED** [review]
The CI/CD workflow mostly looks good. I've added some minor questions to clarify my understanding of how things work.

File: .github/workflows/deploy-production.yml (hunk 15 above)
Comment (tfidfwastaken) [reply .github/workflows/deploy-production.yml:29]: The `PROD_KNOWN_HOSTS` secret is grabbed from `ssh-keyscan -t rsa <REMOTE_HOST_NAME>`, which gets the public key of our deployment server in `known_hosts` format.

This lets us verify the authenticity of remote host while bypassing the prompt that usually comes up when connecting to ssh interactively for the first time.

File: deploy.yml (hunk 18 above)
Comment (tfidfwastaken) [reply deploy.yml:10]: Docker Compose automatically [creates a runtime bind mount](https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose) from the secrets file to the `/run/secrets/<secret-name>` path.

File: deploy.yml (hunk 19 above)
Comment (tfidfwastaken) [reply deploy.yml:40]: They are manually copied to the server on initial setup. Like you said, this needs to be documented. Added it as a task for this project.

**Pull request merged**